src/Functions/parseDateTime.cpp

```diff
@@ -165,9 +159,11 @@ namespace
         (res) = *result;                                  \
 }
 
-    template <ErrorHandling error_handling>
+    template <ErrorHandling error_handling, ReturnType return_type>
     struct DateTime
     {
+        static constexpr Int32 min_year = return_type == ReturnType::DateTime64 ? 1900 : 1970;
```

Comment:
Function `parseDateTime*` returns `DateTime` or `DateTime64`. Class name `DateTime` is incorrect. Kindly rename to `ParsedValue` (yes, that's not strictly related to your PR but nevertheless useful).